pkg/indexerpoc/task.go

```diff
@@ -0,0 +1,449 @@
+		return fmt.Errorf("committing transaction: %w", err)
+	}
+
+	slog.InfoContext(ctx, "indexed blocks",
+		"contract", t.contract.GetName(),
+		"network", t.state.NetworkName,
+		"start", startBlock,
+		"end", endBlock,
+		"logs", len(logs),
+	)
+
+	return nil
+}
+
+// nextBatch retrieves a batch of blocks and logs while checking for reorgs.
+func (t *task) nextBatch(
+	ctx context.Context,
+	startBlock, endBlock uint64,
+) ([]types.Log, bool, error) {
+	// Special case - if we have no previous state, no reorg check needed
+	if t.state.BlockNumber == 0 || t.state.BlockHash == (common.Hash{}) {
+		logs, err := t.src.GetLogs(ctx, startBlock, endBlock, t.filter)
+		return logs, false, err
+	}
+
+	// We need to check for reorgs first by verifying the parent hash of the first block
+	firstBlock, err := t.src.GetBlockByNumber(ctx, startBlock)
+	if err != nil {
+		return nil, false, fmt.Errorf("getting first block %d for reorg check on %s: %w",
+			startBlock, t.state.NetworkName, err)
+	}
+
+	// Check if the parent hash matches our stored hash.
+	if firstBlock.ParentHash() != t.state.BlockHash {
+		// Reorg detected!
+		return nil, true, nil
+	}
+
+	// No reorg, retrieve logs
+	// For small ranges or if concurrency is 1, use a simple approach
+	if t.concurrency <= 1 || endBlock-startBlock < 10 {
+		logs, err := t.src.GetLogs(ctx, startBlock, endBlock, t.filter)
+		if err != nil {
+			return nil, false, fmt.Errorf("getting logs on %s: %w", t.state.NetworkName, err)
+		}
+		return logs, false, nil
+	}
+
+	// For larger ranges, use concurrent processing
+	return t.getLogs(ctx, startBlock, endBlock)
+}
+
+// getLogs retrieves logs in parallel for better performance with large batches.
+// This is likely going to be used only when backfilling.
+func (t *task) getLogs(
+	ctx context.Context,
+	startBlock, endBlock uint64,
+) ([]types.Log, bool, error) {
+	var (
+		allLogs []types.Log
+		mu      sync.Mutex
+		wg      sync.WaitGroup
+		errCh   = make(chan error, t.concurrency)
+	)
+
+	totalBlocks := endBlock - startBlock + 1
+	blocksPerWorker := totalBlocks / uint64(t.concurrency)
+	if blocksPerWorker == 0 {
+		blocksPerWorker = 1
+	}
+
+	for i := 0; i < t.concurrency; i++ {
+		wg.Add(1)
+
+		workerStart := startBlock + uint64(i)*blocksPerWorker
+		workerEnd := workerStart + blocksPerWorker - 1
+
+		// Adjust the last worker to include any remainder
+		if i == t.concurrency-1 {
+			workerEnd = endBlock
+		}
+
+		// Skip if we've already gone past the end
+		if workerStart > endBlock {
+			wg.Done()
+			continue
+		}
+
+		go func(start, end uint64) {
+			defer wg.Done()
+
+			logs, err := t.src.GetLogs(ctx, start, end, t.filter)
+			if err != nil {
+				errCh <- fmt.Errorf("worker getting logs %d-%d on %s: %w",
+					start, end, t.state.NetworkName, err)
+				return
+			}
+
+			mu.Lock()
+			allLogs = append(allLogs, logs...)
+			mu.Unlock()
+		}(workerStart, workerEnd)
+	}
+
+	// Wait for all workers
+	wg.Wait()
+	close(errCh)
+
+	// Check for errors
```

Comment:
@coderabbitai is this the right way to find the errors? Is it correct when using a closed channel?